src/main/java/org/opensearch/knn/index/codec/util/KNNCodecUtil.java

```diff
@@ -17,6 +17,14 @@ public class KNNCodecUtil {
 
     public static final String HNSW_EXTENSION = ".hnsw";
     public static final String HNSW_COMPOUND_EXTENSION = ".hnswc";
+    // Floats are 4 bytes in size
+    public static final int FLOAT_BYTE_SIZE = 4;
+    // References to objects are 4 bytes in size
+    public static final int JAVA_REFERENCE_SIZE = 4;
+    // Each array in Java has a header that is 12 bytes
+    public static final int JAVA_ARRAY_HEADER_SIZE = 12;
```

Comment:
@martin-gaievski I think you removed this header for vectors. Can you please provide some details on if we need this header bytes or not.

Reply:
If the method takes `float[][]` argument as a parameter, wouldn't that guarantee that it would be receiving an array as an input?

Reply:
I think I could use the `getSerializationModeFromStream` method to determine if it's an array or a collection.  I could then modify the `calculateArraySize` method accordingly

Reply:
@navneet1v @martin-gaievski The only way that I can get the SerializationMode to distinguish the type of storage would be to modify the Pair inner class [here](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/codec/util/KNNCodecUtil.java) to add an instance variable.  I would then modify the `getFloats` method in that same class using `serializerModeFromStream` within the try block.  Given that this happens for every merge/refresh operation, would this significantly increase latency?

Reply:
@ryanbogan how much does this array header add as an overhead? if we remove it what would be the impact?

Reply:
@navneet1v It adds `12 * number of vectors + 12` bytes to the total

Reply:
This can be huge. can we somehow read first byte of the binary value and identify what is the type?

Reply:
That's what the method described above would do.  The `serializerModeFromStream` reads the first few bytes and identifies if it's an array or a collection.  Then, it resets the byte stream and returns which type it is.

Reply:
@navneet1v I just pushed a commit to implement the proposed solution above.  It should be easy to revert if necessary, but I don't see a better way of calculating the size of the data.

Reply:
I think it make sense. Thanks for the changes